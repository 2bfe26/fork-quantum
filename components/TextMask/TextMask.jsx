import React from 'react';
import PropTypes from 'prop-types';
import { conformToMask } from 'react-text-mask';
import TextMaskTypes from './TextMaskTypes';

class TextMask extends React.Component {
  static CEP = TextMaskTypes.CEP;

  static CNPJ = TextMaskTypes.CNPJ;

  static CPF = TextMaskTypes.CPF;

  static Date = TextMaskTypes.Date;

  static Phone = TextMaskTypes.Phone;

  render() {
    const { children, text, mask } = this.props;
    const input = text || children;

    return mask ? conformToMask(input, mask).conformedValue : input;
  }
}

TextMask.defaultProps = {
  text: '',
  children: '',
  mask: false,
};

TextMask.propTypes = {
  /** Value to be masked. If ommited, it will be the children prop. */
  text: PropTypes.string,
  children: PropTypes.string,
  /**
   * Mask must follow this [rules](https://github.com/text-mask/text-mask/blob/master/componentDocumentation.md#mask)
   */
  mask: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.func,
    PropTypes.bool,
    PropTypes.shape({
      mask: PropTypes.oneOfType([PropTypes.array, PropTypes.func]),
      pipe: PropTypes.func,
    }),
  ]),
};

export default TextMask;
