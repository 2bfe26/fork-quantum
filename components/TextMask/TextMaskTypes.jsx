import React from 'react';
import TextMask from './TextMask';
import masks from '../shared/masks';

const TextMaskTypes = {
  CEP: props => <TextMask mask={masks.cep} {...props} />,
  CNPJ: props => <TextMask mask={masks.cnpj} {...props} />,
  CPF: props => <TextMask mask={masks.cpf} {...props} />,
  Date: props => <TextMask mask={masks.date} {...props} />,
  Phone: props => <TextMask mask={masks.phone} {...props} />,
};

function setDisplayName(name) {
  TextMaskTypes[name].displayName = `TextMask.${name}`;
}

Object.keys(TextMaskTypes).map(setDisplayName);

export default TextMaskTypes;
