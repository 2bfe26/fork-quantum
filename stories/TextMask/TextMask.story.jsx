import React from 'react';
import { storiesOf } from '@storybook/react';
import {
  Heading,
  TabbedView,
  Tab,
  AutoPropsApi,
  HowToImport,
  Example,
  CodeExample,
} from '@catho/quantum-storybook-ui';

import TextMask from '../../components/TextMask';
import TextMaskTypes from '../../components/TextMask/TextMaskTypes';

const examples = {
  CEP: {
    code:
      'Alameda Juari, 262 - Barueri, SP - <TextMask.CEP>06460090</TextMask.CEP>',
    component: (
      <React.Fragment>
        Alameda Juari, 262 - Barueri, SP - <TextMask.CEP>06460090</TextMask.CEP>
      </React.Fragment>
    ),
  },
  CPNJ: {
    code: 'CNPJ: <TextMask.CNPJ text="03753088000100" />',
    component: (
      <React.Fragment>
        CNPJ: <TextMask.CNPJ text="03753088000100" />
      </React.Fragment>
    ),
  },
  'Card Number': {
    code: `Card Number:
<TextMask mask={[/\\d/, /\\d/, /\\d/, /\\d/, " ", /\\d/, /\\d/, /\\d/, /\\d/, " ", /\\d/, /\\d/, /\\d/, /\\d/, " ", /\\d/, /\\d/, /\\d/, /\\d/]}>4024007113535149</TextMask>`,
    component: (
      <React.Fragment>
        <TextMask
          mask={[
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
          ]}
        >
          4024007113535149
        </TextMask>
      </React.Fragment>
    ),
  },
  CPF: {
    code: `// works with formated text too!
<TextMask.CPF>301.703.170-56</TextMask.CPF>`,
    component: (
      <React.Fragment>
        <TextMask.CPF>301.703.170-56</TextMask.CPF>
      </React.Fragment>
    ),
  },
};

const TextMaskCode = `<TextMask text="06460090" mask={[[/\\d/, /\\d/, /\\d/, /\\d/, /\\d/, '-', /\\d/, /\\d/, /\\d/]]} />
<TextMask.CEP>06460090</TextMask.CEP>
<TextMask.CEP text="06460090" />`;

storiesOf('7. Mask', module).add('Text', () => (
  <React.Fragment>
    <Heading name="TextMask">
      <TabbedView>
        <Tab title="Usage">
          <HowToImport importModules="TextMask" />

          <p>
            <br />
            This is a component to render properly the given text:
          </p>

          <CodeExample showTitle={false} code={TextMaskCode} />
          <br />

          <p>
            All examples above will renders this output:{' '}
            <TextMask.CEP>06460090</TextMask.CEP>
          </p>

          <p>
            Like Input, you can provide your own mask using <code>mask</code>{' '}
            property or use one of the availables:
          </p>

          <ul>
            {Object.keys(TextMaskTypes).map(type => (
              <li key={type}>
                {type}: <code>{`<TextMast.${type} />`}</code>{' '}
              </li>
            ))}
          </ul>
        </Tab>
        <Tab title="API">
          <AutoPropsApi component={TextMask} />
        </Tab>
        <Tab title="Examples">
          {Object.keys(examples).map(example => {
            const { code, component } = examples[example];
            return (
              <Example
                title={example}
                key={example}
                code={code}
                component={component}
              />
            );
          })}
        </Tab>
      </TabbedView>
    </Heading>
  </React.Fragment>
));
