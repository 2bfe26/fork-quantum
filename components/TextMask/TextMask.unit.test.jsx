import React from 'react';
import { mount, shallow } from 'enzyme';
import TextMask from './TextMask';

describe('TextMask component', () => {
  it('Should match the snapshot', () => {
    expect(shallow(<TextMask />)).toMatchSnapshot();
    expect(shallow(<TextMask.CEP />)).toMatchSnapshot();
    expect(shallow(<TextMask.CNPJ />)).toMatchSnapshot();
    expect(shallow(<TextMask.CPF />)).toMatchSnapshot();
    expect(shallow(<TextMask.Date />)).toMatchSnapshot();
    expect(shallow(<TextMask.Phone />)).toMatchSnapshot();
  });

  describe('Should format text', () => {
    const execStaticMaskTest = ({ mask, original, expected }) => {
      const TextMaskClass = TextMask[mask] || TextMask;

      const withValue = mount(<TextMaskClass text={original} />);
      const withChildren = mount(<TextMaskClass>{original}</TextMaskClass>);

      expect(withValue.text()).toBe(expected);
      expect(withChildren.text()).toBe(expected);
    };

    it('No mask', () => {
      execStaticMaskTest({
        original: '12345',
        expected: '12345',
      });
    });

    it('CEP', () => {
      execStaticMaskTest({
        mask: 'CEP',
        original: '06824980',
        expected: '06824-980',
      });
    });

    it('CNPJ', () => {
      execStaticMaskTest({
        mask: 'CNPJ',
        original: '85818714000164',
        expected: '85.818.714/0001-64',
      });
    });

    it('Date', () => {
      execStaticMaskTest({
        mask: 'Date',
        original: '15101988',
        expected: '15/10/1988',
      });
    });

    it('House Phone', () => {
      execStaticMaskTest({
        mask: 'Phone',
        original: '1144445555',
        expected: '(11) 4444-5555',
      });
    });

    it('Cell Phone', () => {
      execStaticMaskTest({
        mask: 'Phone',
        original: '11999998888',
        expected: '(11) 99999-8888',
      });
    });

    it('Custom maks', () => {
      const original = '5102341298321293';
      const expected = '5102 3412 9832 1293';
      const mask = [/\d/, /\d/, /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/];

      const withValue = mount(<TextMask text={original} mask={mask} />);
      const withChildren = mount(<TextMask mask={mask}>{original}</TextMask>);

      expect(withValue.text()).toBe(expected);
      expect(withChildren.text()).toBe(expected);
    });
  });
});
