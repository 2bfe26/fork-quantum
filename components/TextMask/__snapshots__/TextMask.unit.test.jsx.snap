// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextMask component Should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextMask
    mask={false}
    text=""
  >
    
  </TextMask>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): "",
  Symbol(enzyme.__nodes__): Array [
    "",
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`TextMask component Should match the snapshot 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextMask.CEP />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": "",
      "mask": Array [
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        "-",
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
      ],
      "text": "",
    },
    "ref": null,
    "rendered": "",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "",
        "mask": Array [
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          "-",
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
        ],
        "text": "",
      },
      "ref": null,
      "rendered": "",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`TextMask component Should match the snapshot 3`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextMask.CNPJ />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": "",
      "mask": Array [
        /\\\\d/,
        /\\\\d/,
        ".",
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        ".",
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        "/",
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        "-",
        /\\\\d/,
        /\\\\d/,
      ],
      "text": "",
    },
    "ref": null,
    "rendered": "",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "",
        "mask": Array [
          /\\\\d/,
          /\\\\d/,
          ".",
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          ".",
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          "/",
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          "-",
          /\\\\d/,
          /\\\\d/,
        ],
        "text": "",
      },
      "ref": null,
      "rendered": "",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`TextMask component Should match the snapshot 4`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextMask.CPF />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": "",
      "mask": Array [
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        ".",
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        ".",
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        "-",
        /\\\\d/,
        /\\\\d/,
      ],
      "text": "",
    },
    "ref": null,
    "rendered": "",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "",
        "mask": Array [
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          ".",
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          ".",
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          "-",
          /\\\\d/,
          /\\\\d/,
        ],
        "text": "",
      },
      "ref": null,
      "rendered": "",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`TextMask component Should match the snapshot 5`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextMask.Date />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": "",
      "mask": Array [
        /\\\\d/,
        /\\\\d/,
        "/",
        /\\\\d/,
        /\\\\d/,
        "/",
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
      ],
      "text": "",
    },
    "ref": null,
    "rendered": "",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "",
        "mask": Array [
          /\\\\d/,
          /\\\\d/,
          "/",
          /\\\\d/,
          /\\\\d/,
          "/",
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
        ],
        "text": "",
      },
      "ref": null,
      "rendered": "",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`TextMask component Should match the snapshot 6`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextMask.Phone />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": "",
      "mask": [Function],
      "text": "",
    },
    "ref": null,
    "rendered": "",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": "",
        "mask": [Function],
        "text": "",
      },
      "ref": null,
      "rendered": "",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
